### PR TITLE
Capture and include cuBLAS logs in error messages

### PR DESCRIPTION
**Pull request by andportnoy**
This installs a cuBLAS logger callback to capture cuBLAS logs for the duration of all calls to the cuBLAS API that happen through `CUDABlas::DoBlasInternalImpl`. If an error occurs, the captured logs are included as part of the error message in an attempt to provide more context to the user.

Submitting as draft for now because I need to validate that this doesn't cause a performance regression.

### CODE REVIEW

**Comment by andportnoy** [comment]
Sanity checking on a small Paxml model does not show a change in performance.

**Review by cheshire — COMMENTED** [review]
This seems very useful, but I'm concerned about potential CPU runtime cost? Do we have any benchmarks?

**Comment by cheshire** [comment]
The performance characteristics would probably depend on sizes and number of matmuls; do we have numbers on how much CPU overhead does it add? 

Also CC @ezhulenev , are there any CUDA graphs implications?

**Comment by andportnoy** [comment]
@cheshire Let me know if there's a publicly available model you'd like me to benchmark. I tried a Paxml model and didn't see a change in iters/sec. I can follow up with CPU runtime profiles.

**Comment by cheshire** [comment]
Large paxml training model is probably less sensitive to CPU overhead, this
starts to matter when CPU cannot comfortably run ahead of the GPU.

**Comment by kamaljeeti** [comment]
Hi @andportnoy , @cheshire,  Any update on this PR? Please. Thank you!

**Comment by andportnoy** [comment]
@kamaljeeti I'm now actively working on addressing George's questions.
@cheshire For reference, the Paxml model I tried is the 126M on a single H100. Do you have any suggestions of a workload where the CPU doesn't run ahead of the GPU? I have a couple ideas I need to test but wondering if you know of an existing example.

**Comment by andportnoy** [comment]
The overall effect is that the `GemmThunk` CPU runtime is doubled (~10 µs -> ~20 µs on an H100 DGX). This is hidden when cuBLAS GEMMs are interspersed with other operations, or when they are captured into a CUDA graph.

On the other hand if many GEMMs run consecutively back-to-back, there is an observable degradation if the problem size is small enough and thunk CPU runtime exceeds the GPU kernel runtime. This effect existed previously of course, but with the proposed change the threshold is doubled.

Below I'm showing the before and after the change runtime measurements of a sequence of 100 square float32 cuBLAS GEMMs running back to back. Measurements are taken from the start of the first thunk (CPU) to the end of the last kernel (GPU).
| M=K=N  | Before, ms | After, ms | Change |
| - | - | - | - |
| 256 | 1.058 | 1.978 | +86.96% |
| 512 | 1.154 | 1.978 | +71.40% |
| 1024 | 2.358 | 2.38 | +0.93% |
| 2048 | 6.042 | 6.043 | +0.02% |

The Nsight Systems profiles are available [here](https://drive.google.com/drive/folders/11FFpzv-aY18r4iL82yfzXMl8iASL0rSS). JAX was at commit `fad3e749`, XLA "before" was at `c307686e`, "after" at `47d3dc72` (=`c307686e` + the changes = this branch). I used this JAX script:
```python
import sys
import jax
import jax.numpy as jnp

M = int(sys.argv[1])

ITER = 100
As = [jnp.identity(M) for _ in range(ITER)]
Bs = [jnp.identity(M) for _ in range(ITER)]

@jax.jit
def f(As, Bs):
    return [A@B for A,B in zip(As,Bs)]

f(As, Bs)
```

**Comment by cheshire** [comment]
Right, 10us seem significant enough to notice.

@ezhulenev WDYT? Maybe we can run this only under vmodule? Or only if CUDA graphs are disabled? We can't do this with CUDA graphs in any case?

**Comment by andportnoy** [comment]
After a sync, the suggestion from @cheshire was to enable log capture during the autotuning stage, but disable otherwise. This should allow coverage for the same cuBLAS API calls yet avoid the paying the cost at runtime proper.

File: xla/stream_executor/cuda/cuda_blas.cc [hunk 1]
```diff
@@ -396,11 +489,19 @@ absl::Status CUDABlas::DoBlasInternalImpl(FuncT cublas_func, Stream *stream,
                                            : CUBLAS_POINTER_MODE_DEVICE)) {
     return absl::InternalError("Failed setting error mode");
   }
+
+  if (capture_logs) CublasLogCollector::On();
```
Comment (cheshire) [inline xla/stream_executor/cuda/cuda_blas.cc:493]: Nit: usually we use RAII to make sure log collection is always turned off on all paths from this function

**Review by cheshire — CHANGES_REQUESTED** [review]
Would a second PR enable this from autotuning?

File: xla/stream_executor/cuda/cuda_blas.cc (hunk 1 above)
Comment (andportnoy) [reply xla/stream_executor/cuda/cuda_blas.cc:493]: Do you mean a pattern like this?
```c++
if (capture_logs) {
  CublasLogCollector collector;  // constructor enables log capture
  ...
  // destructor disables log capture
} else {
  ...
}
```

**Comment by andportnoy** [comment]
> Would a second PR enable this from autotuning?

@cheshire I must be missing some context, but log capture _is_ enabled in autotuning here if by autotuning we mean the GEMM algorithm picker pass and any other pass that runs GEMMs with `profile_output != nullptr`, which is exactly when we turn log collection on.

File: xla/stream_executor/cuda/cuda_blas.cc (hunk 1 above)
Comment (andportnoy) [reply xla/stream_executor/cuda/cuda_blas.cc]: Implemented `ScopedCublasLogCollection` in a new commit.

**Review by cheshire — CHANGES_REQUESTED** [review]
Could you clarify how autotuner sets the right parameter?

**Comment by kamaljeeti** [comment]
Hi @andportnoy, can you have a look at the comments? Thanks.

**Comment by andportnoy** [comment]
Hi @kamaljeeti, I offered the following opinion to @cheshire offline:

>I'm realizing now the autotuning stage might not be the right place for this. There are two issues:
> 1. Currently errors and corresponding error messages are discarded under the autotuner (i.e. when a ProfileResult output pointer is passed to the GEMM wrapper functions), since the strategy is to try an algorithm and simply exclude it from consideration if it fails: https://github.com/openxla/xla/blob/675da4b2d57c91d623439e0a41c89e1e36bd9c93/xla/stream_executor/blas.h#L149-L167 So trying to catch errors and provide error messages to the user during autotuning goes against the autotuning strategy.
> 2. If the autotuner evaluates multiple algorithms, some of which fail and return errors but some don't, then we don't want to print the logs for each failing algorithm. When all algorithms fail, even that isn't fatal because there is fallback to a default algorithm.

George didn't seem to agree with this. @nouiz and @olupton suggested moving this error messaging mode to the first (only) invocation of the thunk. George is OOTO and we are planning to continue the discussion then.

**Comment by kamaljeeti** [comment]
Hi @andportnoy, can you please update and keep us posted? Thanks.

**Comment by andportnoy** [comment]
I'm withdrawing the PR. We think that including this machinery only during autotuning adds complexity without solving the problem we were hoping to address — providing good error messages for cuBLAS API call errors during model execution, as opposed to during compilation. The idea was to anticipate such failures, but we didn't have a specific example in mind. Maybe when we see a concrete failure in the future, we'll come up with a more elegant solution at that time :)

**Pull request closed without merging**